pkg/operator/watchdog/cmd.go

```diff
@@ -0,0 +1,247 @@
+	fs.StringVar(&o.KubeConfig, "kubeconfig", o.KubeConfig, "kubeconfig file or empty")
+	fs.StringVar(&o.Namespace, "namespace", o.Namespace, "namespace to report the watchdog events")
+	fs.DurationVar(&o.Interval, "interval", 5*time.Second, "interval specify how aggressive the file checks should be")
+}
+
+func (o *FileWatcherOptions) Complete() error {
+	clientConfig, err := client.GetKubeConfigOrInClusterConfig(o.KubeConfig, nil)
+	if err != nil {
+		return err
+	}
+	kubeClient, err := kubernetes.NewForConfig(clientConfig)
+	if err != nil {
+		return err
+	}
+
+	ctx, cancel := context.WithTimeout(context.TODO(), 10*time.Second)
+	defer cancel()
+
+	// Get event recorder.
+	// Retry on connection errors, but don't error out, instead fallback to the namespace.
+	var eventTarget *v1.ObjectReference
+	err = retry.RetryOnConnectionErrors(ctx, func(context.Context) (bool, error) {
+		var clientErr error
+		eventTarget, clientErr = events.GetControllerReferenceForCurrentPod(kubeClient, o.Namespace, nil)
+		if clientErr != nil {
+			return false, clientErr
+		}
+		return true, nil
+	})
+	if err != nil {
+		klog.Warningf("unable to get owner reference (falling back to namespace): %v", err)
+	}
+	o.recorder = events.NewRecorder(kubeClient.CoreV1().Events(o.Namespace), "file-change-watchdog", eventTarget)
+
+	return nil
+}
+
+func (o *FileWatcherOptions) Validate() error {
+	if len(o.ProcessName) == 0 {
+		return fmt.Errorf("process name must be specified")
+	}
+	if len(o.Files) == 0 {
+		return fmt.Errorf("at least one file to observe must be specified")
+	}
+	if len(o.Namespace) == 0 && len(os.Getenv("POD_NAMESPACE")) == 0 {
+		return fmt.Errorf("either namespace flag or POD_NAMESPACE environment variable must be specified")
+	}
+	return nil
+}
+
+// startPidObserver runs a loop that observe changes to the PID of the process we send signals after change is detected.
+// When the PID is observed for the first time, the pidObservedCh is closed.
+// When the PID change is observed (after we got initial PID), the shutdown is called (to signal watchdog to restart).
+func (o *FileWatcherOptions) startPidObserver(shutdown context.CancelFunc, pidObservedCh chan int) {
+	defer shutdown()
+	defer close(pidObservedCh)
+	currentPID := 0
+	for {
+		// get the monitored process PID
+		observedPID, found, err := o.getPIDByNameFn(o.ProcessName)
+		if !found {
+			klog.Warningf("Process %q does not have PID", o.ProcessName)
+			return
+		}
+		if err != nil {
+			klog.Warning(err)
+			return
+		}
+
+		if currentPID == 0 {
+			currentPID = observedPID
+			pidObservedCh <- observedPID
+		}
+
+		// when the PID change, signal shutdown
+		if currentPID != observedPID {
+			return
+		}
+
+		time.Sleep(1 * time.Second)
+	}
+}
+
+// readInitialFileContent reads the content of files specified.
+// This is needed by file observer.
+func (o *FileWatcherOptions) readInitialFileContent() error {
```

Comment:
how do we make sure that this is the state the operand sees? We can be late, and miss an update.

Reply:
If we restart, yes. Practically I don't see a big chance to miss a change (we rotate certs on hour basis). IF this will be monitoring configs, then yes.

We can discuss restart vs. monitor via Slack.

Reply:
lets go with the monitor for now (no crashing)